provider/internal/queue/provide.go

```diff
@@ -0,0 +1,191 @@
+	subtrie, _ := keyspace.FindSubtrie(q.keys, prefix)
+	keys := keyspace.AllValues(subtrie, bit256.ZeroKey())
+
+	// Remove the keys from the keys trie.
+	keyspace.PruneSubtrie(q.keys, prefix)
+
+	return prefix, keys, true
+}
+
+// DequeueMatching returns keys matching the given prefix from the queue.
+//
+// The keys and prefix are removed from the queue. If the queue is empty, or
+// supplied prefix doesn't match any keys, an empty slice is returned.
+func (q *ProvideQueue) DequeueMatching(prefix bitstr.Key) []mh.Multihash {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	subtrie, ok := keyspace.FindSubtrie(q.keys, prefix)
+	if !ok {
+		// No keys matching the prefix.
+		return nil
+	}
+	keys := keyspace.AllValues(subtrie, bit256.ZeroKey())
+
+	// Remove the keys from the keys trie.
+	keyspace.PruneSubtrie(q.keys, prefix)
+
+	// Remove prefix and its superstrings from queue if any.
+	removed := q.queue.Remove(prefix)
+	if !removed {
+		// prefix and superstrings not in queue.
+		if shorterPrefix, ok := keyspace.FindPrefixOfKey(q.queue.prefixes, prefix); ok {
+			// prefix is a superstring of some other shorter prefix in the queue.
+			// Leave it in the queue, unless the shorter prefix doesn't have any
+			// matching keys left.
+			if _, ok := keyspace.FindSubtrie(q.keys, shorterPrefix); !ok {
+				// No keys matching shorterPrefix, remove shorterPrefix from queue.
+				q.queue.Remove(shorterPrefix)
+			}
+		}
+	}
+	return keys
+}
+
+// Remove removes the supplied keys from the queue.
+//
+// If this operation removes the last keys for prefixes in the queue, remove
+// the prefixes from the queue.
+func (q *ProvideQueue) Remove(keys ...mh.Multihash) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	matchingPrefixes := make(map[bitstr.Key]struct{})
+
+	// Remove keys from the keys trie.
+	for _, h := range keys {
+		k := keyspace.MhToBit256(h)
+		q.keys.Remove(k)
+		if prefix, ok := keyspace.FindPrefixOfKey(q.queue.prefixes, k); ok {
+			// Get the trie leaf matching the key, if any.
+			matchingPrefixes[prefix] = struct{}{}
+		}
+	}
+
+	// For matching prefixes, if no more keys are matching, remove them from
+	// queue.
+	prefixesToRemove := make([]bitstr.Key, 0)
+	for prefix := range matchingPrefixes {
+		if _, ok := keyspace.FindSubtrie(q.keys, prefix); !ok {
+			prefixesToRemove = append(prefixesToRemove, prefix)
+		}
+	}
+	if len(prefixesToRemove) > 0 {
+		q.queue.removePrefixesFromQueue(prefixesToRemove)
+	}
+}
+
+// IsEmpty returns true if the queue is empty.
+func (q *ProvideQueue) IsEmpty() bool {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.queue.Size() == 0
+}
+
+// Size returns the number of regions containing at least one key in the queue.
+func (q *ProvideQueue) Size() int {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.sizeNoLock()
+}
+
+// sizeNoLock returns the number of regions containing at least one key in the
+// queue. It assumes the mutex is held already.
+func (q *ProvideQueue) sizeNoLock() int {
+	return q.keys.Size()
+}
```

Comment:
This seems like an unnecessary function. Would it be better to just call `q.keys.Size()`?